File: 59-spiral-matrix-ii/59-spiral-matrix-ii.java
class Solution {
    public int[][] generateMatrix(int n) {
        int[][] ans = new int[n][n];
        int[][] ds = {{0,1},{1,0},{0,-1},{-1,0}};

        int d = 0;
        int r = 0; int c = 0;
        for (int i = 1; i < (n*n+1); i++) {
            ans[r][c] = i;


            if (r + ds[d][0] < 0 || r + ds[d][0] >= n || c + ds[d][1] < 0 || c+ds[d][1] >=n || ans[r+ds[d][0]][c+ds[d][1]] != 0) {
                d = (d+1)%4;
            };
            r += ds[d][0];
            c += ds[d][1];
            
        };
        
        return ans;
        
        
    }
}

File: 59-spiral-matrix-ii/59-spiral-matrix-ii.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        matrix = [[0]*n for i in range(n)]
        
        
        d = 0
        pos = (0, 0)
        for num in range(1, n*n+1):

            currR, currC = pos
    
            matrix[currR][currC] = num
            
            dR, dC = directions[d%4]
            pos = (currR + dR, currC + dC) 
            if pos[0] < 0 or pos[1] < 0 or pos[0] >= n or pos[1] >= n:
                d += 1
            elif matrix[pos[0]][pos[1]] != 0:
                d += 1
            dR, dC = directions[d%4]
            pos = (currR + dR, currC + dC) 
            
        return matrix
                        
            